### PR TITLE
Load global symbols

### PR DESCRIPTION
**Pull request by ahcorde**
This is the basic PR to make use of `ign-plugin` in other libraries.

This comment in this other PR https://github.com/ignitionrobotics/ign-sensors/pull/38#issuecomment-678123125 explains what is going on. Basically some libraries pre-load some symbols of other libraries (for example `libignition-gazebo4-sensors-system.so` pre-load some ign-sensor symbols) and when you need to load some of the ign-sensor libraries you can't access these symbols because it was load previously.

Using `RTLD_GLOBAL` we can access theses symbols but we should filter in all the libraries and look for the right one.

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

**Comment by traversaro** [comment]
Just as a curiosity, which version of dlfcn-win32 is used in the CI ? vcpkg has 1.1.1 while conda-forge has 1.2.0, and the 1.2.0 had some changes w.r.t. 1.1.1 related also to the RTLD_GLOBAL flag (see https://github.com/dlfcn-win32/dlfcn-win32/pull/44).

File: loader/src/Loader.cc [hunk 1]
```diff
@@ -452,9 +452,8 @@ namespace ignition
       // state gets cleared each time it is called.
       dlerror();
 
-      // NOTE: We open using RTLD_LOCAL instead of RTLD_GLOBAL to prevent the
-      // symbols of different libraries from writing over each other.
-      void *dlHandle = dlopen(_full_path.c_str(), RTLD_LAZY | RTLD_LOCAL);
+      // NOTE: We are using RTLD_GLOBAL, this may overwrite the symbols of different libraries.
```
Comment (chapulina) [inline loader/src/Loader.cc:455]: `./loader/src/Loader.cc:455:  Lines should be <= 80 characters long  [whitespace/line_length] [2]`

File: loader/src/Loader.cc [hunk 2]
```diff
@@ -536,12 +535,6 @@ namespace ignition
         const std::string& _pathToLibrary) const
     {
       std::vector<Info> loadedPlugins;
-
-      // This function should never be called with a nullptr _dlHandle
-      assert(_dlHandle &&
-             "Bug in code: Loader::Implementation::LoadPlugins was called with "
-             "a nullptr value for _dlHandle.");
```
Comment (chapulina) [inline loader/src/Loader.cc:543]: Isn't this safety check needed anymore?

**Review by chapulina — COMMENTED** [review]
Do you think it would be feasible to write a test case that fails without this PR?

File: loader/src/Loader.cc (hunk 2 above)
Comment (ahcorde) [reply loader/src/Loader.cc:543]: We should call dlopen with `nullptr` to get all the symbols from the executable. According with the [documentation](https://linux.die.net/man/3/dlopen):

```txt
If filename is NULL, then the returned handle is for the main program
```

Comment (chapulina) [reply loader/src/Loader.cc:543]: But `_dlHandle` is being used in `dlsym` below, not `dlopen` :monocle_face:

Comment (ahcorde) [reply loader/src/Loader.cc:543]: Sorry, this is the right one:

```
If dlsym() is called with the special handle NULL, it is interpreted as a reference to the executable or shared object from which the call is being made. Thus a shared object can reference its own symbols.
```

Comment (chapulina) [reply loader/src/Loader.cc:543]: This sounds like a potential change in behaviour, but I haven't looked closely at this yet.

I noticed you removed the Blueprint / Citadel / Dome labels from this PR and left Edifice. This PR is targeting `ign-plugin1`, so it would apply for all those versions though. Is your intention to target a new major version?

Comment (ahcorde) [reply loader/src/Loader.cc:543]: There aren't any changes in the API, I think we can target the rest of the versions. This will require to backport the changes in all the "LoadPlugin" libraries for all the versions

**Comment by ahcorde** [comment]
Thinking a little bit more about this I think this should target `main`. There are no changes in the API but the *behaviour* of the library has changed and this will required to modify the existing code that makes use of ign-plugin1. What do you think @mjcarroll and @chapulina ?

**Comment by mjcarroll** [comment]
I agree that `main` probably makes more sense.  @chapulina are we bumping plugin for Edifice?

**Comment by chapulina** [comment]
> are we bumping plugin for Edifice?

If you peeps are certain that this change is needed to get `ign-sensors` to work, I'm ok bumping `ign-plugin` in Edifice. Every downstream library is already being bumped, so this shouldn't have a large impact.

**Comment by ahcorde** [comment]
Retargeted to `main`.

**Comment by ahcorde** [comment]
friendly ping @mjcarroll and @chapulina

**Comment by chapulina** [comment]
> Just as a curiosity, which version of dlfcn-win32 is used in the CI ?

That's a question for @j-rivero

File: test/integration/EnablePluginFromThis_TEST.cc [hunk 3]
```diff
@@ -136,8 +134,6 @@ TEST(EnablePluginFromThis, LibraryManagement)
   }
 
   EXPECT_TRUE(weak.IsExpired());
-
-  CHECK_FOR_LIBRARY(libraryPath, false);
```
Comment (chapulina) [inline test/integration/EnablePluginFromThis_TEST.cc:140]: I believe these checks are here to make sure the library is unloaded correctly. But @mxgrey's comment on the function suggests that it may not be reliable, so that could explain why you had to remove these. What do you think @mxgrey , should we remove the false checks, or adjust the `CHECK_FOR_LIBRARY` function somehow?

https://github.com/ignitionrobotics/ign-plugin/blob/07c20cbc61007bf0982583b3eecb7761ca5cf7c9/test/integration/utils.hh#L32-L42

File: test/integration/plugin.cc [hunk 4]
```diff
@@ -574,8 +540,25 @@ TEST(PluginPtr, LibraryManagement)
 
     CHECK_FOR_LIBRARY(path, true);
   }
+}
+
+/////////////////////////////////////////////////
+TEST(Loader, LoadBadPlugins)
```
Comment (chapulina) [inline test/integration/plugin.cc:546]: Was there a reason to move this test down here? Was it failing above?

File: loader/src/Loader.cc [hunk 5]
```diff
@@ -368,7 +368,7 @@ namespace ignition
 #endif
 
       void *dlHandle = dlopen(_pathToLibrary.c_str(),
-                              RTLD_NOLOAD | RTLD_LAZY | RTLD_LOCAL);
+                              RTLD_NOLOAD | RTLD_LAZY | RTLD_GLOBAL);
```
Comment (chapulina) [inline loader/src/Loader.cc:371]: What do you think, @mxgrey , are we trying to do something that `ign-plugin` wasn't meant to do, or is this a reasonable change?

See https://github.com/ignitionrobotics/ign-sensors/pull/38#issuecomment-678123125 for the reason why this change is needed.

**Comment by j-rivero** [comment]
> > Just as a curiosity, which version of dlfcn-win32 is used in the CI ?
> 
> That's a question for @j-rivero

Uhm. We are using tarball in the ign-plugin build, I can not find the version on the code. The date of the tarball is Nov 2017, probably 1.1.1?

**Comment by chapulina** [comment]
I talked to @mxgrey and he had the nice suggestion of adding a flag to enable the new behaviour. If we do this, we may not need to bump the version and only enable the new behaviour for the libraries that need it.

Then I started looking more closely and noticed that the `dlopen` call being changed is inside `ForgetLibrary`. I hadn't noticed this detail before. I think it explains the changes in the tests. But now I'm confused about how this helps `ign-sensors`. Can you elaborate a bit more on why this is needed, @ahcorde ?

**Comment by ahcorde** [comment]
@chapulina 

Adding a flag required to keep the two functionalities at the same time but the idea it's to keep just one way to load libraries (remove the code in ign-common if I remember correctly).

The function `dlclose()` decrements the reference count on the dynamic library handle. If the reference count drops to zero and no other loaded libraries use symbols in it, then the dynamic library is unloaded. But this is not reliable. Anyhow I'm also using `NODELETE` to avoid unload something that it's already running. I remember I got some segfault without this .

I not really sure about how this could help the different libraries (ign-sensors, ign-gui, ign-rendering or ign-gazebo). But the idea it's to use `ign-plugin` which is the library focuses on loading shared libraries. Maybe there is a deeper implementation issue inside `ign-plugin` and we don't want to move forward with this implemention, because we nned to load all the symbols an filter them in each library.

**Comment by traversaro** [comment]
> 
> 
> > > Just as a curiosity, which version of dlfcn-win32 is used in the CI ?
> > 
> > 
> > That's a question for @j-rivero
> 
> Uhm. We are using tarball in the ign-plugin build, I can not find the version on the code. The date of the tarball is Nov 2017, probably 1.1.1?

Thanks! I am asking as in the journey to get a reliable working Gazebo with vcpkg-installed dependencies I became the dlfcn-win32 maintainer, even if I am definitively not an expert of Windows shared library interface. However, several users over time contributed quite a lot to the library. In particular in the release 1.2.0 that came out in 2019 (https://github.com/dlfcn-win32/dlfcn-win32/releases/tag/v1.2.0) there have been some works on loading of global symbols (see https://github.com/dlfcn-win32/dlfcn-win32/pull/44). I would expect for everything to work fine for both dlfcn-win32 1.1.1 and 1.2.0, but with this tricky issues problems can always emerge.

**Comment by chapulina** [comment]
Removing beta label, we won't have time to wrap this up before code freeze. Let's retarget at Ignition-F.

**Comment by chapulina** [comment]
This isn't necessary anymore as we solved the `ign-sensors` needs by removing plugin functionality.
